src/middleware/axum.rs

```diff
@@ -0,0 +1,136 @@
+// which is licensed under CC0 1.0 Universal
+// https://github.com/davidB/tracing-opentelemetry-instrumentation-sdk/blob/d3609ac2cc699d3a24fbf89754053cc8e938e3bf/LICENSE
+
+use axum::extract::MatchedPath;
+use http::{Request, Response};
+use pin_project_lite::pin_project;
+use std::{
+    error::Error,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+use tracing::Span;
+use tracing_opentelemetry_instrumentation_sdk::http as otel_http;
+
+pub type Filter = fn(&str) -> bool;
+
+#[derive(Default, Debug, Clone)]
+pub struct OtelAxumLayer {
+    filter: Option<Filter>,
+}
+
+// add a builder like api
+impl OtelAxumLayer {
+    #[must_use]
+    pub fn filter(self, filter: Filter) -> Self {
+        OtelAxumLayer {
+            filter: Some(filter),
+        }
+    }
+}
+
+impl<S> Layer<S> for OtelAxumLayer {
+    /// The wrapped service
+    type Service = OtelAxumService<S>;
+    fn layer(&self, inner: S) -> Self::Service {
+        OtelAxumService {
+            inner,
+            filter: self.filter,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct OtelAxumService<S> {
+    inner: S,
+    filter: Option<Filter>,
+}
+
+impl<S, B, B2> Service<Request<B>> for OtelAxumService<S>
+where
+    S: Service<Request<B>, Response = Response<B2>> + Clone + Send + 'static,
+    S::Error: Error + 'static, //fmt::Display + 'static,
+    S::Future: Send + 'static,
+    B: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    // #[allow(clippy::type_complexity)]
+    // type Future = futures_core::future::BoxFuture<'static, Result<Self::Response, Self::Error>>;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<B>) -> Self::Future {
+        use tracing_opentelemetry::OpenTelemetrySpanExt;
+        let span = if self.filter.map_or(true, |f| f(req.uri().path())) {
+            let span = otel_http::http_server::make_span_from_request(&req);
+            let route = http_route(&req);
+            let method = otel_http::http_method(req.method());
+            // let client_ip = parse_x_forwarded_for(req.headers())
+            //     .or_else(|| {
+            //         req.extensions()
+            //             .get::<ConnectInfo<SocketAddr>>()
+            //             .map(|ConnectInfo(client_ip)| Cow::from(client_ip.to_string()))
+            //     })
+            //     .unwrap_or_default();
+            span.record("http.route", route);
+            span.record("otel.name", format!("{method} {route}").trim());
+            // span.record("trace_id", find_trace_id_from_tracing(&span));
+            // span.record("client.address", client_ip);
```

Comment:
Older version extracted a lot of information from `axum` request:

```rust
        let span = tracing::info_span!(
            "HTTP request",
            otel.name= %name,
            http.client_ip = %client_ip,
            http.flavor = %http_flavor(req.version()),
            http.host = %host,
            http.method = %http_method_v,
            http.route = %http_route,
            http.scheme = %scheme,
            http.status_code = Empty,
            http.target = %http_target,
            http.user_agent = %user_agent,
            otel.kind = %"server", //opentelemetry::trace::SpanKind::Server
            otel.status_code = Empty,
            trace_id = %trace_id,
        );
```

It looks like the latest version doesn't extract the same metadata.

Reply:
I figured it out.

``rust
let span = otel_http::http_server::make_span_from_request(&req);
```

does this now.

Reply:
So, `trace_id` and `http.client.address` are the only fields missing now.

But perhaps it's OK.

Reply:
yes, I am resoliving this conversation for now